### PR TITLE
gitserver: use an atomic alternative to os.RemoveAll

### PR DESCRIPTION
**Pull request by ijt**
Fixes https://github.com/sourcegraph/sourcegraph/issues/3972

Test plan: This PR includes a unit test, including a sub-test for atomicity.

### CODE REVIEW

File: cmd/gitserver/server/cleanup.go [hunk 1]
```diff
@@ -628,3 +628,18 @@ func removeFileOlderThan(path string, maxAge time.Duration) error {
 	}
 	return nil
 }
+
+func removeAllAtomically(path string) error {
+	if _, err := os.Stat(path); err != nil {
+		return nil
+	}
+	d, err := ioutil.TempDir("", "removeAtomically")
```
Comment (keegancsmith) [inline cmd/gitserver/server/cleanup.go]: this assumes `$TMPDIR`  is on the same mount as path, otherwise `os.Rename` fails / is not atomic.

**Comment by tsenart** [comment]
I'll skip review since @keegancsmith already took a look at it.

File: cmd/gitserver/server/cleanup.go (hunk 1 above)
Comment (ijt) [reply cmd/gitserver/server/cleanup.go]: Ah, that would explain why it's done that way in `Server.removeRepoDirectory`.

**Comment by ijt** [comment]
> I'm unsure why you don't want to use removeRepoDir. It is well tested and is how we remove all other repository directories. See the inline feedback for an example bug in your implementation.
> 
> Note: We use a specific tmp dir since we also have logic to clean up that temp directory in a performant way if the process dies while deleting. IE the logic in removeRepoDirectory has come from the battle hardening gitserver.

It seemed overly complicated and mixed concerns of atomic dir deletion with removal of empty ancestors. I see your point. Opened a new PR for this: https://github.com/sourcegraph/sourcegraph/pull/4006.

**Pull request closed without merging**